### PR TITLE
[PHN-1049] Use KafkaJS instead of node-rdkafka

### PR DESCRIPTION
**Pull request by giulliano-bueno**
- Remove node-rdkafka
- Add KafkaJS

Most of the changes were because KafkaJS has a different implementation and does not have a lot of events used by node-rdkafka (on ready, error and etc).

### CODE REVIEW

File: src/kafka-pubsub.ts [hunk 1]
```diff
@@ -39,213 +40,129 @@ export class KafkaPubSub extends PubSubEngine {
...
-          } else {
-            resolve()
-          }
-        })
-      }
-    })
-    let consumerPromise: Promise<void>
-    consumerPromise = new Promise((resolve, reject) => {
-      if (this.consumer) {
-        this.consumer.disconnect((err) => {
```
Comment (mohaazeem) [inline src/kafka-pubsub.ts:121]: Why is this removed?

**Comment by mohaazeem** [comment]
I have a question: how do we plan to keep this up-to-date with the main repo? Or do we not need their support?

File: src/kafka-pubsub.ts [hunk 2]
```diff
@@ -39,213 +40,129 @@ export class KafkaPubSub extends PubSubEngine {
...
-            {},
-            {},
-            this.options.topicConfig
-        )       
-    );
-    producer.on('event.error', (err) => {
-      this.logger.error(err)
-    })
-    return new Promise((resolve, reject) => {
-      producer.on('ready', (data, metadata) => {
```
Comment (rodrigopinto) [inline src/kafka-pubsub.ts:163]: I believe we should keep the instrumentation for the connect. Maybe it will to avoid issues when trying to produce before the producer being connected. https://kafka.js.org/docs/instrumentation-events#a-name-producer-a-producer

File: src/kafka-pubsub.ts [hunk 3]
```diff
@@ -39,213 +40,129 @@ export class KafkaPubSub extends PubSubEngine {
...
-        let topics =  metadata.topics.map(topic => topic.name);
-        this.logger.info('Connected, found topics: %s', topics);
-        
-        if (topics.includes(this.options.topic)) {
-          resolve(producer);
-        } else {
-          this.logger.error('Could not find requested topic %s', this.options.topic);
-          producer.disconnect()
```
Comment (rodrigopinto) [inline src/kafka-pubsub.ts:171]: Ditto.

File: src/kafka-pubsub.ts [hunk 4]
```diff
@@ -39,213 +40,129 @@ export class KafkaPubSub extends PubSubEngine {
...
+    await producer.connect()
+    return producer
   }
 
-  private async createConsumer(topic: string): Promise<Kafka.KafkaConsumer> {    
-    // Create a group for each instance. The consumer will receive all messages from the topic
-    const groupId = this.options.groupId || uuidv4()
+  private async handleMessage(topic: string, partition: number, message: Kafka.Message) {
+    const prefix = `${topic}[${partition}] / ${message.timestamp}`
+    console.log(`- ${prefix} ${message.key}#${message.value}`)
```
Comment (rodrigopinto) [inline src/kafka-pubsub.ts:132]: this console log I think can be removed.

File: src/kafka-pubsub.ts [hunk 5]
```diff
@@ -39,213 +40,129 @@ export class KafkaPubSub extends PubSubEngine {
...
+  }
 
-    })
-    
-    consumer.on('event.log', (event) => {
-      this.logger.debug(event);
-    });
-
-    return new Promise((resolve, reject) => {
-      consumer.on('ready', (data, metadata) => {
```
Comment (rodrigopinto) [inline src/kafka-pubsub.ts:231]: Same for the consumer.

File: src/kafka-pubsub.ts [hunk 6]
```diff
@@ -39,213 +40,129 @@ export class KafkaPubSub extends PubSubEngine {
...
-        this.logger.info('Connected, found topics: %s', topics);
-        
-        if (topics.includes(topic)) {
-          this.logger.info("Subscribing to %s", topic)
-          consumer.subscribe([topic]);
-          consumer.consume();
-          resolve(consumer);
-        } else {
-          this.logger.error('Could not find requested topic %s', topic);
-          consumer.disconnect()
```
Comment (rodrigopinto) [inline src/kafka-pubsub.ts:242]: We need to treat the disconnet case.

**Review by rodrigopinto — COMMENTED** [review]
We need to apply few changes before getting this merged.

File: src/kafka-pubsub.ts (hunk 2 above)
Comment (giulliano-bueno) [reply src/kafka-pubsub.ts:163]: The most similar event would be CONNECT. Do you think it would be enough?

Comment (giulliano-bueno) [reply src/kafka-pubsub.ts:163]: But an exception would be generated in case of error (AFAIK).

File: src/kafka-pubsub.ts (hunk 5 above)
Comment (giulliano-bueno) [reply src/kafka-pubsub.ts:231]: Again, the same argument used before.

File: src/kafka-pubsub.ts (hunk 2 above)
Comment (giulliano-bueno) [reply src/kafka-pubsub.ts:163]: The await would be enough here AFAIK.

File: src/kafka-pubsub.ts [hunk 7]
```diff
@@ -25,11 +25,12 @@ const defaultLogger = Logger.createLogger({
 
 export class KafkaPubSub extends PubSubEngine {
 
-  protected producer: Kafka.HighLevelProducer // ProducerStream not exported
-  protected consumer: Kafka.KafkaConsumer // ConsumerStream not exported 
+  protected kafka: Kafka.Kafka
+  protected producer: Kafka.Producer // ProducerStream not exported
```
Comment (giulliano-bueno) [inline src/kafka-pubsub.ts:29]: TODO: check if this producer will do a round robin (partition).

File: src/kafka-pubsub.ts (hunk 1 above)
Comment (giulliano-bueno) [reply src/kafka-pubsub.ts:121]: Good catch! I forgot to add the consumer.

File: src/kafka-pubsub.ts [hunk 8]
```diff
@@ -39,213 +40,129 @@ export class KafkaPubSub extends PubSubEngine {
...
-    const groupId = this.options.groupId || uuidv4()
+  private async handleMessage(topic: string, partition: number, message: Kafka.Message) {
+    const prefix = `${topic}[${partition}] / ${message.timestamp}`
+    console.log(`- ${prefix} ${message.key}#${message.value}`)
 
-    const consumer = new Kafka.KafkaConsumer(
-        Object.assign(
-          {},
-          {
-            'group.id': `kafka-pubsub-${groupId}`,
```
Comment (giulliano-bueno) [inline src/kafka-pubsub.ts:189]: This is not possible for us since our authentication would fail if the informed group id is not exactly the same as configured in our authentication.

File: src/kafka-pubsub.ts (hunk 6 above)
Comment (giulliano-bueno) [reply src/kafka-pubsub.ts:242]: Indeed. We need to make sure we are not disconnecting and not generating an alert or exception.

File: src/kafka-pubsub.ts (hunk 1 above)
Comment (giulliano-bueno) [reply src/kafka-pubsub.ts:121]: Fixed.

**Review by mohaazeem — APPROVED** [review]
